### PR TITLE
Lesliefang/optimized softmax

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
vectorized softmax calculation when not along the last dim. 
Changes:
1. Use vectorized `softmax_kernel` instead of host_softmax when not along the last dim.
2. Special optimization of bf16 dtype to reduce data convert.

Performance data on 28 cores' CPX:
Input size [32, 81, 15130] and do softmax along the second dim(81).
* FP32 Baseline: 24.67 ms
* FP32 optimized: 9.2 ms
* BF16 optimized: 6.13 ms

### CODE REVIEW

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp [hunk 1]
```diff
@@ -206,6 +207,268 @@ struct vec_host_softmax_lastdim {
...
+    int64_t dim_size) { // for example: 81
+  using Vec = vec::Vectorized<scalar_t>;
+  int64_t grain_size = internal::GRAIN_SIZE / (16 * dim_size);
+  if (grain_size < 1)
+    grain_size = 1;
+
+  int64_t dim_stride = inner_size; // for example: 15130
+  int64_t outer_stride = dim_size * dim_stride; // for example: 81 * 15130
+
+  parallel_for(
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/SoftMaxKernel.cpp:338]: outer_size and inner_size are not adjacent dimensions, we should not collapse non-adjacent dimensions, otherwise it will result into non-contiguous memory access.

Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:338]: How about we only parallel_for along the `inner_size`? And there will be a outer for loop(serial) along the `outer_size`?

**Comment by leslie-fang-intel** [comment]
todo:
1. Remove the outsize in the parallel for
2. For BF16, we shouldn't allocate https://github.com/leslie-fang-intel/pytorch/pull/3/files#diff-1d3fef70dbc64e3e716d056f635ed159f3b3bb161930bc126317c5ff11d88957R351-R352 instead, we should do the load/store bf16/fp32 convert in each of 3 steps of softmax.
